packages/@aws-cdk/aws-stepfunctions-tasks/lib/call-dynamodb.ts

```diff
@@ -0,0 +1,791 @@
+  }
+
+  /**
+   * Sets an attribute of type Number. For example:  "N": "123.45"
+   * Numbers are sent across the network to DynamoDB as strings,
+   * to maximize compatibility across languages and libraries.
+   * However, DynamoDB treats them as number type attributes for mathematical operations.
+   */
+  public withN(value: string) {
+    this.attributeValue.N = value;
+    return this;
+  }
+
+  /**
+   * Sets an attribute of type Binary. For example:  "B": "dGhpcyB0ZXh0IGlzIGJhc2U2NC1lbmNvZGVk"
+   */
+  public withB(value: string) {
+    this.attributeValue.B = value;
+    return this;
+  }
+
+  /**
+   * Sets an attribute of type String Set. For example:  "SS": ["Giraffe", "Hippo" ,"Zebra"]
+   */
+  public withSS(value: string[]) {
+    this.attributeValue.SS = value;
+    return this;
+  }
+
+  /**
+   * Sets an attribute of type Number Set. For example:  "NS": ["42.2", "-19", "7.5", "3.14"]
+   * Numbers are sent across the network to DynamoDB as strings,
+   * to maximize compatibility across languages and libraries.
+   * However, DynamoDB treats them as number type attributes for mathematical operations.
+   */
+  public withNS(value: string[]) {
+    this.attributeValue.NS = value;
+    return this;
+  }
+
+  /**
+   * Sets an attribute of type Binary Set. For example:  "BS": ["U3Vubnk=", "UmFpbnk=", "U25vd3k="]
+   */
+  public withBS(value: string[]) {
+    this.attributeValue.BS = value;
+    return this;
+  }
+
+  /**
+   * Sets an attribute of type Map. For example:  "M": {"Name": {"S": "Joe"}, "Age": {"N": "35"}}
+   */
+  public withM(value: DynamoAttributeValueMap) {
+    this.attributeValue.M = transformAttributeValueMap(value);
+    return this;
+  }
+
+  /**
+   * Sets an attribute of type List. For example:  "L": [ {"S": "Cookies"} , {"S": "Coffee"}, {"N", "3.14159"}]
+   */
+  public withL(value: DynamoAttributeValue[]) {
+    this.attributeValue.L = value.map(val => val.toObject());
+    return this;
+  }
+
+  /**
+   * Sets an attribute of type Null. For example:  "NULL": true
+   */
+  public withNULL(value: boolean) {
+    this.attributeValue.NULL = value;
+    return this;
+  }
+
+  /**
+   * Sets an attribute of type Boolean. For example:  "BOOL": true
+   */
+  public withBOOL(value: boolean) {
+    this.attributeValue.BOOL = value;
+    return this;
+  }
+
+  /**
+   * Return the attributeValue object
+   */
+  public toObject() {
+    return this.attributeValue;
+  }
+}
+
+/**
+ * Property for any key
+ */
+export interface DynamoAttribute {
+  /**
+   * The name of the attribute
+   */
+  readonly name: string;
+
+  /**
+   * The value of the attribute
+   */
+  readonly value: DynamoAttributeValue;
+}
+
+/**
+ * Class to generate projection expression
+ */
+export class DynamoProjectionExpression {
+  private expression: string[] = [];
+
+  /**
+   * Adds the passed attribute to the chain
+   *
+   * @param attr Attribute name
+   */
+  public withAttribute(attr: string): DynamoProjectionExpression {
```

Comment:
Same here around supporting tokens. Make sure this works correctly and a test.

Reply:
I am not sure how we would properly handle this scenario with tokens since we are doing string concatenation here.
For:
```ts
projectionExpression: [
        new tasks.DynamoProjectionExpression()
          .withAttribute('Messages')
          .atIndex(1)
          .withAttribute(sfn.Data.stringAt('$.tagsKey')),
        new tasks.DynamoProjectionExpression().withAttribute(
          sfn.Data.stringAt('$.idKey')
        )
      ]
```

I am getting the following error:
```
Field references must be the entire string, cannot concatenate them (found 'Messages[1].${Token[tagsKey.19]},${Token[idKey.20]}')
```

Reply:
> 'Messages[1].${Token[tagsKey.19]},${Token[idKey.20]}'

This should be valid. Where is this error coming from?

Reply:
This is the complete stack trace of the above error

```
Field references must be the entire string, cannot concatenate them (found 'Messages[1].${Token[tagKey.21]},${Token[otherKey.22]}')

      197 | 
      198 |   if (jsonPathTokens.length > 0 && fragments.length > 1) {
    > 199 |     throw new Error(`Field references must be the entire string, cannot concatenate them (found '${x}')`);
          |           ^
      200 |   }
      201 |   if (jsonPathTokens.length > 0) {
      202 |     return jsonPathTokens[0].path;

      at jsonPathString (../aws-stepfunctions/lib/json-path.ts:199:11)
      at Object.renderString (../aws-stepfunctions/lib/json-path.ts:146:16)
      at recurseObject (../aws-stepfunctions/lib/json-path.ts:90:35)
      at Object.renderObject (../aws-stepfunctions/lib/json-path.ts:36:10)
      at Function.renderObject (../aws-stepfunctions/lib/fields.ts:111:12)
      at Task.toStateJson (../aws-stepfunctions/lib/states/task.ts:161:65)
      at Object.<anonymous> (test/call-dynamodb.test.ts:251:29)
```

Reply:
Not entirely sure what's going on here. This is good enough for now. 
This will hopefully get fixed up in a later PR.

What you have here is already going to benefit a ton of customers!